File: Arrays and Strings/permutation/permutation.cpp
#include <iostream>
#include <map>
#include <string>

using namespace std;

bool permutation(const string &a, const string &b) {
  if (a.size() != b.size()) return false;
  map<char, int> counts;
  for (char c : a) ++counts[c];
  for (char c : b) {
    --counts[c];
    if (counts[c] < 0) return false;
  }
  return true;
}

int main() {
  string a, b;
  cout << "Enter a string: ";
  cin >> a;
  cout << "Enter another string: ";
  cin >> b;
  cout << (permutation(a, b) ? "The strings are permuations of each other"
	   : "The strings are not permuations of each other") << endl;
}

File: Arrays and Strings/permutation/Permutation.java
import java.util.*;

class Permutation {

    public static boolean permuation(String a, String b) {
	if (a.length() != b.length()) return false;
	HashMap<Character, Integer> counts = new HashMap<Character, Integer>();
	    
	char[] achars = a.toCharArray();
	char[] bchars = b.toCharArray();

	for (char c : achars) {
	    if (counts.get(c) == null)
		counts.put(c, 1);
	    else
		counts.put(c, counts.get(c) + 1);
	}
	for (char c : bchars) {
	    if (counts.get(c) == null) return false;
	    counts.put(c, counts.get(c) - 1);
	    if (counts.get(c) < 0) return false;
	}

	return true;

    }

    public static void main(String[] args) {
	Scanner sc = new Scanner(System.in);
	System.out.print("Enter a string: ");
	String a = sc.next();
	System.out.print("Enter another string: ");
	String b = sc.next();
	if (permuation(a, b))
	    System.out.println("The strings are permuations of each other");
	else
	    System.out.println("The strings are not permutations of each other");
		
    }

}
